src/function.ts

```diff
@@ -18,10 +18,8 @@ export interface RustFunctionProps extends FunctionOptions {
 
   /**
    * The Lambda runtime to deploy this function with. `provided.al2023` is the default.
-   *
-   * The only valid values are `provided.al2023` and `provided.al2`.
    */
-  readonly runtime?: string;
+  readonly runtime?: 'provided.al2023' | 'provided.al2';
```

Comment:
To allow the user all available runtimes, it could also be typed like this: `readonly runtime?: 'provided.al2023' | 'provided.al2' | Runtime;` and at runtime we would check what was actually provided:
```ts
const runtime = typeof props?.runtime === 'string' ?
      new Runtime(props.runtime) :
      props?.runtime ?
        props.runtime :
        new Runtime('provided.al2023');
```

Reply:
Hmm, that is a weird one. Running `npx projen test` on macOS locally gives this for me:

```
❯ npx projen test 
👾 test | jest --passWithNoTests --updateSnapshot
watchman warning:  Recrawled this watch 19 times, most recently because:
MustScanSubDirs UserDroppedTo resolve, please review the information on
https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl
To clear this warning, run:
`watchman watch-del '<redacted>/cargo-lambda-cdk' ; watchman watch-project '<redacted>cargo-lambda-cdk'`

ts-jest[ts-jest-transformer] (WARN) Define `ts-jest` config under `globals` is deprecated. Please do
transform: {
    <transform_regex>: ['ts-jest', { /* ts-jest config goes here in Jest */ }],
},
ts-jest[ts-jest-transformer] (WARN) Define `ts-jest` config under `globals` is deprecated. Please do
transform: {
    <transform_regex>: ['ts-jest', { /* ts-jest config goes here in Jest */ }],
},
ts-jest[ts-jest-transformer] (WARN) Define `ts-jest` config under `globals` is deprecated. Please do
transform: {
    <transform_regex>: ['ts-jest', { /* ts-jest config goes here in Jest */ }],
},
ts-jest[ts-jest-transformer] (WARN) Define `ts-jest` config under `globals` is deprecated. Please do
transform: {
    <transform_regex>: ['ts-jest', { /* ts-jest config goes here in Jest */ }],
},
 PASS  test/cargo.test.ts
Bundling asset Default/rust extension/Code/Stage...
Bundling asset Default/arm-64-set-via-lambda-architecture/Code/Stage...
Bundling asset Default/rust function/Code/Stage...
    Finished release [optimized] target(s) in 0.01s
    Finished release [optimized] target(s) in 0.00s
Bundling asset Default/rust function 1/Code/Stage...
    Finished release [optimized] target(s) in 0.00s
Bundling asset Default/rust extension 1/Code/Stage...
Bundling asset Default/arm-64-set-via-lambda-architecture/Code/Stage...
warning: some crates are on edition 2021 which defaults to `resolver = "2"`, but virtual workspaces default to `resolver = "1"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
    Finished release [optimized] target(s) in 0.00s
Bundling asset Default/rust extension 2/Code/Stage...
warning: some crates are on edition 2021 which defaults to `resolver = "2"`, but virtual workspaces default to `resolver = "1"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
    Finished release [optimized] target(s) in 0.00s
Bundling asset Default/rust function 2/Code/Stage...
    Finished release [optimized] target(s) in 0.00s
Bundling asset Default/arm-64-set-via-lambda-architecture/Code/Stage...
warning: some crates are on edition 2021 which defaults to `resolver = "2"`, but virtual workspaces default to `resolver = "1"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
note: to use the edition 2021 resolver, specify `workspace.resolver = "2"` in the workspace root's manifest
    Finished release [optimized] target(s) in 0.00s
warning: some crates are on edition 2021 which defaults to `resolver = "2"`, but virtual workspaces default to `resolver = "1"`
note: to keep the current resolver, specify `workspace.resolver = "1"` in the workspace root's manifest
"` in the workspace root's manifestver, specify `workspace.resolver = "2
release [optimized] target(s) in 0.00s
Bundling asset Default/rust function/Code/Stage...
 PASS  test/extension.test.ts (6.41 s)
    Finished release [optimized] target(s) in 0.00s
 PASS  test/bundlingOptions.test.ts (6.462 s)
    Finished release [optimized] target(s) in 0.00s
 PASS  test/function.test.ts (6.994 s)
--------------|---------|----------|---------|---------|--------------------------------
File          | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s              
--------------|---------|----------|---------|---------|--------------------------------
All files     |   87.28 |    73.15 |   93.75 |   86.95 |                                
 bundling.ts  |   83.07 |    67.64 |   88.88 |   82.53 | 75,131-132,185,190-196,201,229 
 cargo.ts     |     100 |      100 |     100 |     100 |                                
 extension.ts |     100 |       75 |     100 |     100 | 38-46                          
 function.ts  |     100 |    77.77 |     100 |     100 | 48-60                          
 index.ts     |     100 |      100 |     100 |     100 |                                
 util.ts      |      75 |     72.6 |     100 |      75 | 13,17-20                       
--------------|---------|----------|---------|---------|--------------------------------

Test Suites: 4 passed, 4 total
Tests:       20 passed, 20 total
Snapshots:   0 total
Time:        7.223 s
Ran all test suites.
👾 test » eslint | eslint --ext .ts,.tsx --fix --no-error-on-unmatched-pattern src test build-tools .projenrc.js
=============

WARNING: You are currently running a version of TypeScript which is not officially supported by @typescript-eslint/typescript-estree.

You may find that it works just fine, or you may not.

SUPPORTED TYPESCRIPT VERSIONS: >=3.3.1 <5.2.0

YOUR TYPESCRIPT VERSION: 5.2.2

Please only submit bug reports when using the officially supported version.

=============
```

But now that we talk about it: I think we can't use this type at all because it isn't supported in all jsii supported languages. I don't think that e.g. go or Java support a type like this. I looked over https://aws.github.io/jsii/user-guides/lib-author/typescript-restrictions/ but couldn't find an example that resembles our problem. What do you think?